File: docs/api_examples.py
"""
FastBlog API Python 客户端示例

本文件提供完整的 API 调用示例，支持同步和异步调用
需要安装: pip install httpx aiohttp
"""

from typing import Optional, Dict, Any


# ============================================================================
# 同步客户端 (使用 httpx)
# ============================================================================

class FastBlogClient:
    """FastBlog API 同步客户端"""

    def __init__(self, base_url: str = "http://localhost:9421/api/v1"):
        self.base_url = base_url
        self.access_token: Optional[str] = None
        self.session = None

    def _get_headers(self) -> Dict[str, str]:
        """获取请求头"""
        headers = {"Content-Type": "application/json"}
        if self.access_token:
            headers["Authorization"] = f"Bearer {self.access_token}"
        return headers

    def _request(self, method: str, endpoint: str, **kwargs) -> Dict[str, Any]:
        """发送 HTTP 请求"""
        import httpx

        url = f"{self.base_url}{endpoint}"

        with httpx.Client() as client:
            response = client.request(
                method=method,
                url=url,
                headers=self._get_headers(),
                **kwargs
            )
            response.raise_for_status()
            return response.json()

    # 🔐 认证 Auth

    def login(self, username: str, password: str) -> Dict[str, Any]:
        """用户登录"""
        data = self._request("POST", "/auth/login", json={
            "username": username,
            "password": password
        })
        self.access_token = data.get("access_token")
        print("✅ Login successful")
        return data

    def refresh_token(self) -> Dict[str, Any]:
        """刷新 Token"""
        return self._request("POST", "/auth/refresh")

    def logout(self) -> None:
        """登出"""
        self._request("POST", "/auth/logout")
        self.access_token = None
        print("✅ Logout successful")

    # 📝 文章 Articles

    def get_articles(
            self,
            page: int = 1,
            per_page: int = 10,
            search: Optional[str] = None,
            category_id: Optional[int] = None,
            user_id: Optional[int] = None,
            status: Optional[str] = None
    ) -> Dict[str, Any]:
        """获取文章列表"""
        params = {"page": page, "per_page": per_page}
        if search:
            params["search"] = search
        if category_id:
            params["category_id"] = category_id
        if user_id:
            params["user_id"] = user_id
        if status:
            params["status"] = status

        return self._request("GET", "/articles", params=params)

    def get_article(self, article_id: int) -> Dict[str, Any]:
        """获取文章详情"""
        return self._request("GET", f"/articles/{article_id}")

    def create_article(self, article: Dict[str, Any]) -> Dict[str, Any]:
        """创建文章"""
        return self._request("POST", "/articles", json=article)

    def update_article(self, article_id: int, updates: Dict[str, Any]) -> Dict[str, Any]:
        """更新文章"""
        return self._request("PUT", f"/articles/{article_id}", json=updates)

    def delete_article(self, article_id: int) -> Dict[str, Any]:
        """删除文章"""
        return self._request("DELETE", f"/articles/{article_id}")

    # 📂 分类 Categories

    def get_categories(self) -> Dict[str, Any]:
        """获取分类列表"""
        return self._request("GET", "/categories")

    def create_category(self, category: Dict[str, Any]) -> Dict[str, Any]:
        """创建分类"""
        return self._request("POST", "/categories", json=category)

    # 🖼️ 媒体 Media

    def upload_file(self, file_path: str, folder: str = "uploads") -> Dict[str, Any]:
        """上传文件"""
        with open(file_path, "rb") as f:
            files = {"file": f}
            data = {"folder": folder}
            return self._request("POST", "/media/upload", files=files, data=data)

    def get_media(self, page: int = 1, per_page: int = 20) -> Dict[str, Any]:
        """获取媒体列表"""
        return self._request("GET", "/media", params={"page": page, "per_page": per_page})

    # 👥 用户 Users

    def get_current_user(self) -> Dict[str, Any]:
        """获取当前用户信息"""
        return self._request("GET", "/users/me")

    def get_users(self, page: int = 1, per_page: int = 10) -> Dict[str, Any]:
        """获取用户列表"""
        return self._request("GET", "/users", params={"page": page, "per_page": per_page})

    # 💬 评论 Comments

    def get_comments(self, article_id: int, page: int = 1, per_page: int = 20) -> Dict[str, Any]:
        """获取文章评论"""
        return self._request("GET", "/comments", params={
            "article_id": article_id,
            "page": page,
            "per_page": per_page
        })

    def create_comment(self, comment: Dict[str, Any]) -> Dict[str, Any]:
        """发表评论"""
        return self._request("POST", "/comments", json=comment)

    # 📊 仪表板 Dashboard

    def get_dashboard_stats(self) -> Dict[str, Any]:
        """获取统计数据"""
        return self._request("GET", "/dashboard/stats")

    def get_dashboard_analytics(self, days: int = 30) -> Dict[str, Any]:
        """获取分析数据"""
        return self._request("GET", "/dashboard/analytics", params={"days": days})

    # 🔌 插件 Plugins

    def get_plugins(self) -> Dict[str, Any]:
        """获取插件列表"""
        return self._request("GET", "/plugins")

    def activate_plugin(self, slug: str) -> Dict[str, Any]:
        """激活插件"""
        return self._request("POST", f"/plugins/{slug}/activate")

    def deactivate_plugin(self, slug: str) -> Dict[str, Any]:
        """停用插件"""
        return self._request("POST", f"/plugins/{slug}/deactivate")

    # 🎨 主题 Themes

    def get_themes(self) -> Dict[str, Any]:
        """获取主题列表"""
        return self._request("GET", "/themes")

    def activate_theme(self, slug: str) -> Dict[str, Any]:
        """激活主题"""
        return self._request("POST", f"/themes/{slug}/activate")

    # ⚙️ 设置 Settings

    def get_settings(self) -> Dict[str, Any]:
        """获取系统设置"""
        return self._request("GET", "/settings")

    def update_settings(self, updates: Dict[str, Any]) -> Dict[str, Any]:
        """更新设置"""
        return self._request("PUT", "/settings", json=updates)

    # 🤖 AI 功能

    def generate_metadata(
            self,
            title: str,
            content: str,
            excerpt: Optional[str] = None
    ) -> Dict[str, Any]:
        """生成 AI 友好的元数据"""
        return self._request("POST", "/ai/metadata/generate", json={
            "title": title,
            "content": content,
            "excerpt": excerpt
        })


# ============================================================================
# 异步客户端 (使用 aiohttp)
# ============================================================================

class AsyncFastBlogClient:
    """FastBlog API 异步客户端"""

    def __init__(self, base_url: str = "http://localhost:9421/api/v1"):
        self.base_url = base_url
        self.access_token: Optional[str] = None

    async def _get_headers(self) -> Dict[str, str]:
        """获取请求头"""

        headers = {"Content-Type": "application/json"}
        if self.access_token:
            headers["Authorization"] = f"Bearer {self.access_token}"
        return headers

    async def _request(self, method: str, endpoint: str, **kwargs) -> Dict[str, Any]:
        """发送 HTTP 请求"""
        import aiohttp

        url = f"{self.base_url}{endpoint}"

        async with aiohttp.ClientSession() as session:
            async with session.request(
                    method=method,
                    url=url,
                    headers=await self._get_headers(),
                    **kwargs
            ) as response:
                response.raise_for_status()
                return await response.json()

    # 🔐 认证 Auth

    async def login(self, username: str, password: str) -> Dict[str, Any]:
        """用户登录"""
        data = await self._request("POST", "/auth/login", json={
            "username": username,
            "password": password
        })
        self.access_token = data.get("access_token")
        print("✅ Login successful")
        return data

    async def logout(self) -> None:
        """登出"""
        await self._request("POST", "/auth/logout")
        self.access_token = None
        print("✅ Logout successful")

    # 📝 文章 Articles

    async def get_articles(self, **params) -> Dict[str, Any]:
        """获取文章列表"""
        return await self._request("GET", "/articles", params=params)

    async def get_article(self, article_id: int) -> Dict[str, Any]:
        """获取文章详情"""
        return await self._request("GET", f"/articles/{article_id}")

    async def create_article(self, article: Dict[str, Any]) -> Dict[str, Any]:
        """创建文章"""
        return await self._request("POST", "/articles", json=article)

    async def update_article(self, article_id: int, updates: Dict[str, Any]) -> Dict[str, Any]:
        """更新文章"""
        return await self._request("PUT", f"/articles/{article_id}", json=updates)

    async def delete_article(self, article_id: int) -> Dict[str, Any]:
        """删除文章"""
        return await self._request("DELETE", f"/articles/{article_id}")

    # 📊 仪表板 Dashboard

    async def get_dashboard_stats(self) -> Dict[str, Any]:
        """获取统计数据"""
        return await self._request("GET", "/dashboard/stats")


# ============================================================================
# 使用示例
# ============================================================================

def sync_example():
    """同步客户端使用示例"""
    print("=" * 60)
    print("同步客户端示例")
    print("=" * 60)

    # 创建客户端
    client = FastBlogClient()

    try:
        # 1. 登录
        print("\n📝 Logging in...")
        client.login("admin@example.com", "your_password")

        # 2. 获取文章列表
        print("\n📚 Fetching articles...")
        articles = client.get_articles(page=1, per_page=5)
        print(f"Found {len(articles['data'])} articles")

        # 3. 创建新文章
        print("\n✍️ Creating new article...")
        new_article = client.create_article({
            "title": "My First Article",
            "slug": "my-first-article",
            "excerpt": "This is a test article",
            "content": "# Hello World\n\nThis is my first article using the API!",
            "category_id": 1,
            "tags": ["Test", "API"],
            "status": "draft"
        })
        print(f"Article created with ID: {new_article['data']['id']}")

        # 4. 获取文章详情
        print("\n📖 Fetching article details...")
        article = client.get_article(new_article['data']['id'])
        print(f"Article title: {article['data']['title']}")

        # 5. 更新文章
        print("\n🔄 Updating article...")
        client.update_article(new_article['data']['id'], {
            "title": "Updated Title",
            "status": "published"
        })
        print("Article updated")

        # 6. 获取统计数据
        print("\n📊 Fetching dashboard stats...")
        stats = client.get_dashboard_stats()
        print(f"Total articles: {stats['data']['total_articles']}")

        # 7. 登出
        print("\n👋 Logging out...")
        client.logout()

        print("\n✅ Workflow completed successfully!")

    except Exception as e:
        print(f"\n❌ Workflow failed: {e}")


async def async_example():
    """异步客户端使用示例"""
    print("=" * 60)
    print("异步客户端示例")
    print("=" * 60)

    # 创建客户端
    client = AsyncFastBlogClient()

    try:
        # 1. 登录
        print("\n📝 Logging in...")
        await client.login("admin@example.com", "your_password")

        # 2. 获取文章列表
        print("\n📚 Fetching articles...")
        articles = await client.get_articles(page=1, per_page=5)
        print(f"Found {len(articles['data'])} articles")

        # 3. 获取统计数据
        print("\n📊 Fetching dashboard stats...")
        stats = await client.get_dashboard_stats()
        print(f"Total articles: {stats['data']['total_articles']}")

        # 4. 登出
        print("\n👋 Logging out...")
        await client.logout()

        print("\n✅ Async workflow completed successfully!")

    except Exception as e:
        print(f"\n❌ Async workflow failed: {e}")


if __name__ == "__main__":
    # 运行同步示例
    sync_example()

    # 运行异步示例
    import asyncio

    asyncio.run(async_example())

File: docs/api_examples.js
/**
 * FastBlog API JavaScript 客户端示例
 *
 * 本文件提供完整的 API 调用示例，可直接在浏览器或 Node.js 环境中运行
 */

// ============================================================================
// 配置
// ============================================================================
const BASE_URL = 'http://localhost:9421/api/v1';
let accessToken = '';

// ============================================================================
// 工具函数
// ============================================================================

/**
 * 发送 HTTP 请求
 */
async function request(endpoint, options = {}) {
    const url = `${BASE_URL}${endpoint}`;
    const config = {
        headers: {
            'Content-Type': 'application/json',
            ...options.headers,
        },
        ...options,
    };

    // 添加认证 token
    if (accessToken) {
        config.headers['Authorization'] = `Bearer ${accessToken}`;
    }

    try {
        const response = await fetch(url, config);
        const data = await response.json();

        if (!response.ok) {
            throw new Error(data.error || `HTTP error! status: ${response.status}`);
        }

        return data;
    } catch (error) {
        console.error(`API Request Failed: ${endpoint}`, error);
        throw error;
    }
}

// ============================================================================
// 🔐 认证 Auth
// ============================================================================

/**
 * 用户登录
 */
async function login(username, password) {
    const data = await request('/auth/login', {
        method: 'POST',
        body: JSON.stringify({username, password}),
    });

    accessToken = data.access_token;
    console.log('✅ Login successful');
    return data;
}

/**
 * 刷新 Token
 */
async function refreshToken() {
    return await request('/auth/refresh', {
        method: 'POST',
    });
}

/**
 * 登出
 */
async function logout() {
    await request('/auth/logout', {
        method: 'POST',
    });
    accessToken = '';
    console.log('✅ Logout successful');
}

// ============================================================================
// 📝 文章 Articles
// ============================================================================

/**
 * 获取文章列表
 */
async function getArticles(params = {}) {
    const {page = 1, perPage = 10, search, categoryId, userId, status} = params;

    const queryParams = new URLSearchParams({
        page: page.toString(),
        per_page: perPage.toString(),
    });

    if (search) queryParams.append('search', search);
    if (categoryId) queryParams.append('category_id', categoryId.toString());
    if (userId) queryParams.append('user_id', userId.toString());
    if (status) queryParams.append('status', status);

    return await request(`/articles?${queryParams}`);
}

/**
 * 获取文章详情
 */
async function getArticle(id) {
    return await request(`/articles/${id}`);
}

/**
 * 创建文章
 */
async function createArticle(article) {
    return await request('/articles', {
        method: 'POST',
        body: JSON.stringify(article),
    });
}

/**
 * 更新文章
 */
async function updateArticle(id, updates) {
    return await request(`/articles/${id}`, {
        method: 'PUT',
        body: JSON.stringify(updates),
    });
}

/**
 * 删除文章
 */
async function deleteArticle(id) {
    return await request(`/articles/${id}`, {
        method: 'DELETE',
    });
}

// ============================================================================
// 📂 分类 Categories
// ============================================================================

/**
 * 获取分类列表
 */
async function getCategories() {
    return await request('/categories');
}

/**
 * 创建分类
 */
async function createCategory(category) {
    return await request('/categories', {
        method: 'POST',
        body: JSON.stringify(category),
    });
}

// ============================================================================
// 🖼️ 媒体 Media
// ============================================================================

/**
 * 上传文件
 */
async function uploadFile(file, folder = 'uploads') {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('folder', folder);

    const response = await fetch(`${BASE_URL}/media/upload`, {
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${accessToken}`,
        },
        body: formData,
    });

    return await response.json();
}

/**
 * 获取媒体列表
 */
async function getMedia(page = 1, perPage = 20) {
    return await request(`/media?page=${page}&per_page=${perPage}`);
}

// ============================================================================
// 👥 用户 Users
// ============================================================================

/**
 * 获取当前用户信息
 */
async function getCurrentUser() {
    return await request('/users/me');
}

/**
 * 获取用户列表
 */
async function getUsers(page = 1, perPage = 10) {
    return await request(`/users?page=${page}&per_page=${perPage}`);
}

// ============================================================================
// 💬 评论 Comments
// ============================================================================

/**
 * 获取文章评论
 */
async function getComments(articleId, page = 1, perPage = 20) {
    return await request(`/comments?article_id=${articleId}&page=${page}&per_page=${perPage}`);
}

/**
 * 发表评论
 */
async function createComment(comment) {
    return await request('/comments', {
        method: 'POST',
        body: JSON.stringify(comment),
    });
}

// ============================================================================
// 📊 仪表板 Dashboard
// ============================================================================

/**
 * 获取统计数据
 */
async function getDashboardStats() {
    return await request('/dashboard/stats');
}

/**
 * 获取分析数据
 */
async function getDashboardAnalytics(days = 30) {
    return await request(`/dashboard/analytics?days=${days}`);
}

// ============================================================================
// 🔌 插件 Plugins
// ============================================================================

/**
 * 获取插件列表
 */
async function getPlugins() {
    return await request('/plugins');
}

/**
 * 激活插件
 */
async function activatePlugin(slug) {
    return await request(`/plugins/${slug}/activate`, {
        method: 'POST',
    });
}

/**
 * 停用插件
 */
async function deactivatePlugin(slug) {
    return await request(`/plugins/${slug}/deactivate`, {
        method: 'POST',
    });
}

// ============================================================================
// 🎨 主题 Themes
// ============================================================================

/**
 * 获取主题列表
 */
async function getThemes() {
    return await request('/themes');
}

/**
 * 激活主题
 */
async function activateTheme(slug) {
    return await request(`/themes/${slug}/activate`, {
        method: 'POST',
    });
}

// ============================================================================
// ⚙️ 设置 Settings
// ============================================================================

/**
 * 获取系统设置
 */
async function getSettings() {
    return await request('/settings');
}

/**
 * 更新设置
 */
async function updateSettings(updates) {
    return await request('/settings', {
        method: 'PUT',
        body: JSON.stringify(updates),
    });
}

// ============================================================================
// 🤖 AI 功能
// ============================================================================

/**
 * 生成 AI 友好的元数据
 */
async function generateMetadata(title, content, excerpt = null) {
    return await request('/ai/metadata/generate', {
        method: 'POST',
        body: JSON.stringify({title, content, excerpt}),
    });
}

// ============================================================================
// 使用示例
// ============================================================================

/**
 * 完整工作流程示例
 */
async function exampleWorkflow() {
    try {
        // 1. 登录
        console.log('📝 Logging in...');
        await login('admin@example.com', 'your_password');

        // 2. 获取文章列表
        console.log('\n📚 Fetching articles...');
        const articles = await getArticles({page: 1, perPage: 5});
        console.log(`Found ${articles.data.length} articles`);

        // 3. 创建新文章
        console.log('\n✍️ Creating new article...');
        const newArticle = await createArticle({
            title: 'My First Article',
            slug: 'my-first-article',
            excerpt: 'This is a test article',
            content: '# Hello World\n\nThis is my first article using the API!',
            category_id: 1,
            tags: ['Test', 'API'],
            status: 'draft',
        });
        console.log(`Article created with ID: ${newArticle.data.id}`);

        // 4. 获取文章详情
        console.log('\n📖 Fetching article details...');
        const article = await getArticle(newArticle.data.id);
        console.log(`Article title: ${article.data.title}`);

        // 5. 更新文章
        console.log('\n🔄 Updating article...');
        await updateArticle(newArticle.data.id, {
            title: 'Updated Title',
            status: 'published',
        });
        console.log('Article updated');

        // 6. 获取统计数据
        console.log('\n📊 Fetching dashboard stats...');
        const stats = await getDashboardStats();
        console.log(`Total articles: ${stats.data.total_articles}`);

        // 7. 登出
        console.log('\n👋 Logging out...');
        await logout();

        console.log('\n✅ Workflow completed successfully!');
    } catch (error) {
        console.error('❌ Workflow failed:', error.message);
    }
}

// ============================================================================
// 导出（Node.js 环境）
// ============================================================================
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        // Auth
        login,
        refreshToken,
        logout,

        // Articles
        getArticles,
        getArticle,
        createArticle,
        updateArticle,
        deleteArticle,

        // Categories
        getCategories,
        createCategory,

        // Media
        uploadFile,
        getMedia,

        // Users
        getCurrentUser,
        getUsers,

        // Comments
        getComments,
        createComment,

        // Dashboard
        getDashboardStats,
        getDashboardAnalytics,

        // Plugins
        getPlugins,
        activatePlugin,
        deactivatePlugin,

        // Themes
        getThemes,
        activateTheme,

        // Settings
        getSettings,
        updateSettings,

        // AI
        generateMetadata,

        // Example
        exampleWorkflow,
    };
}

// 在浏览器中直接运行示例
if (typeof window !== 'undefined') {
    console.log('FastBlog API Client loaded');
    console.log('Run exampleWorkflow() to see a complete demo');
}
